Handle 401 unauthorized error in the UI

ChengYanJin opened the pull request:
**Component**: UI

**Context**: 
The aim of this PR is to handle correctly the 401 unauthorize error on the UI side.

**Summary**:
Add Error Reducer in order to dispatch authErrorAction.
We throw out AuthError when we get unauthorized status from Salt API and K8s API, and catch it at the top of reducer.
In the PrivateRoute, we display the `<ErrorPage401/>` if isAuthError is true in the redux-store. 


**Acceptance criteria**: 
EN:
![image](https://user-images.githubusercontent.com/18453133/146037117-5a7a6f2e-2187-41c2-83df-d244e8365b67.png)
FR:
![image](https://user-images.githubusercontent.com/18453133/146161057-4177f73b-0a88-4cf1-b443-6332dacad233.png)

Core-ui v0.25.0 includes this 401 unauthorized error page.

ChengYanJin committed:
ui: Add Error reducer to store the isAuthError

ChengYanJin committed:
ui: Dispatch authErrorAction at the top level of redux

ChengYanJin committed:
ui: Add handleUnAuthorizedError function to throw out Auth error

ChengYanJin committed:
ui/package: Bump the core-ui version to v0.25.0 to have the new 401 page

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.11/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` with contents from `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI`
and `development/2.11`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.11/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI origin/development/2.11
 $ git merge origin/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.11/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI
```

ChengYanJin committed:
Changelog: Handle 401 unauthorized error in MetalK8s UI

Cuervino commented:
@scality/doc Could you please review the wording? 
I'm not sure about the "supplied".

gaudiauj commented on shell-ui/package.json:
```diff
@@ -47,7 +47,7 @@
   },
   "dependencies": {
-    "@scality/core-ui": "github:scality/core-ui.git#v0.22.2.1",
+    "@scality/core-ui": "github:scality/core-ui.git#695c63c",
```
do we really want to keep the commit hash rather than deploying a new version ?

ChengYanJin replied:
Yes. You are right. We should tag a new version for the core-ui. 
I am just waiting a bit to see if I can get a review from @scality/doc about the wording in the 401 unauthorized page.

gaudiauj commented on ui/src/ducks/app/authError.js:
```diff
@@ -0,0 +1,29 @@
+  isAuthError: false,
+};
+
+export default function reducer(
+  state: AuthErrorState = defaultState,
+  action: AuthErrorAction,
+) {
+  switch (action.type) {
+    case 'AUTH_ERROR': {
+      return { ...state, isAuthError: true };
+    }
+    default:
+      return { ...state };
+  }
+}
```
i'm not familiar with the way it work but what i see from the code, there is no way to remove the isAuthError. And we check in `privateRoute.js` if isAuthError is true. Are we sure that if the user relogin the isAuthError is reset ?

ChengYanJin replied:
Hello @gaudiauj ,
It's a very good concern! but thanks to the authentication workflow, the isAuthError should be reset if we re-login again.

Once the user triggers logout action, the page will be redirected to the IDP, either dex or Keycloak. 

Given the page is fully reloaded, the redux store will be re-initiated as well.

gaudiauj reviewed (dismissed):
just couple of questions, but besides that it looks good. 

![approved](https://media.giphy.com/media/3osxYdek8wYWCOLgT6/giphy.gif)

ChengYanJin committed:
shell-ui: Add style to top level container

Following the changing in Errorpage from 100vh to 100%, we have to make
sure the error page takes the remaining entire height

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` into
`development/2.10`
* [w/2.11/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI](https://github.com/scality/metalk8s/compare/development/2.11...w/2.11/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI) will be merged into `development/2.11` 
* [w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

bert-e commented:
# History mismatch



Merge commit #898aaada770194e5d835311af7a9158a6f77fe24 on the integration branch
`w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` is merging a branch which is neither the current
branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

JBWatenbergScality commented on ui/src/services/salt/api.js:
```diff
@@ -113,4 +128,9 @@ export async function getNodesIPsInterfaces(
     ],
   });
+  if (result.error) {
+    handleUnAuthorizedError({ error: result.error });
```
I think it may need to be returned

JBWatenbergScality reviewed (commented):
Other than this comment and Jean's comment LGTM

JBWatenbergScality commented on ui/src/services/salt/api.js:
```diff
@@ -50,17 +57,27 @@ export async function deployNode(node: string, version: string) {
     arg: [jid],
   });
+  if (result.error) {
+    handleUnAuthorizedError({ error: result.error });
```
here too

bert-e commented:
# History mismatch



Merge commit #2432f52b5feba423da595be39b60efbdb66ec818 on the integration branch
`w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` is merging a branch which is neither the current
branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

cathydossantospinto commented:
@ChengYanJin Here's my suggestion for the French part.
 
```
! Accès refusé
Les identifiants fournis ne vous permettent pas de consulter cette page.
Vous pouvez contacter le support pour signaler le problème.
```
@lucieleonard Feel free to suggest something better if you can 😛 

@John-VanHook will review the English part 😉

John-VanHook commented:
The content of the dialog box is not wrong. You could keep it as is. But I suggest this is better: 

You don’t have permission to view this page using the credentials **you have** supplied. 

You **can** contact `support` to report this issue.

ChengYanJin commented:
ping

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# History mismatch



Merge commit #2432f52b5feba423da595be39b60efbdb66ec818 on the integration branch
`w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` is merging a branch which is neither the current
branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` nor the development branch
`development/123.0`.

It is likely due to a rebase of the branch `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

ChengYanJin commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI` into
`development/2.11`
* [w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI](https://github.com/scality/metalk8s/compare/development/123.0...w/123.0/bugfix/ARTESCA-2572-handle-unauthorized-error-in-the-UI) will be merged into `development/123.0` 



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

ChengYanJin commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @gdemonet

* @thomasdanan

* @chengyanjin

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @nicolast

ChengYanJin commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`

* :heavy_check_mark: `development/123.0`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue ARTESCA-2572.

Goodbye chengyanjin.
